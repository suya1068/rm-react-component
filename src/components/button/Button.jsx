import "./Button.scss";
import classnames from "classnames";
import React, { PureComponent } from "react";
import PropTypes from "prop-types";

const hasOwnProperty = Object.prototype.hasOwnProperty;
const toString = Object.prototype.toString;

class Button extends PureComponent {
    render() {
        const { children, theme, size, width, height, className, style, ...rest } = this.props;
        const buttonProps = {
            ...rest,
            className: classnames("u-button", Button.theme[theme], Button.size[size], className)
        };

        if (style || width || height) {
            buttonProps.styles = Object.assign({}, style, Button.createStyles({ width, height }));
        }

        return React.createElement("button", buttonProps, children);
    }
}

Button.theme = {
    "primary": "u-button-primary",
    "secondary": "u-button-secondary",
    "positive": "u-button-positive",
    "negative": "u-button-negative",
    "info": "u-button-info",
    "dark": "u-button-dark"
};

Button.size = {
    "tiny": "u-button-tiny",
    "small": "u-button-small",
    "large": "u-button-large",
    "big": "u-button-big"
};

Button.createStyles = function createStyles(data = {}) {
    const sheet = Object.keys(data).reduce((box, key) => {
        const value = data[key];

        if (value !== undefined && value !== null && value !== "") {
            box[key] = typeof value === "string" ? value : `${value}px`;
        }

        return box;
    }, {});

    return Object.keys(sheet).length > 0 ? sheet : null;
};

Button.propTypes = {
    children: PropTypes.node.isRequired,
    theme: PropTypes.oneOf(Object.keys(Button.theme)),
    size: PropTypes.oneOf(Object.keys(Button.size)),
    width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    className: PropTypes.string,
    style: PropTypes.shape({})
};

Button.defaultProps = {
    theme: null,
    size: null,
    width: null,
    className: "",
    style: null
};

export default Button;
