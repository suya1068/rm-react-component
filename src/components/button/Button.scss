$u-button-line-height: 1.5 !default;
$u-button-padding-x: 0.85rem !default;
$u-button-padding-y: 0.65rem !default;
$u-button-font-size: 1rem !default;

$u-button-line-height-tiny: 1.5 !default;
$u-button-padding-x-tiny: 0.6rem !default;
$u-button-padding-y-tiny: 0.4rem !default;
$u-button-font-size-tiny: 0.75rem !default;

$u-button-line-height-small: 1.5 !default;
$u-button-padding-x-small: 0.725rem !default;
$u-button-padding-y-small: 0.525rem !default;
$u-button-font-size-small: 0.875rem !default;

$u-button-line-height-large: 1.5 !default;
$u-button-padding-x-large: 0.975rem !default;
$u-button-padding-y-large: 0.775rem !default;
$u-button-font-size-large: 1.125rem !default;

$u-button-line-height-big: 1.5 !default;
$u-button-padding-x-big: 1.1rem !default;
$u-button-padding-y-big: 0.9rem !default;
$u-button-font-size-big: 1.25rem !default;

$u-button-border-radius: 2px;
$u-button-box-shadow: 0 0 0 1px transparent inset, 0 0 0 0 rgba(34,36,38,.15) inset;
$u-button-font-family: "Segoe UI", Roboto, "Helvetica Neue", Arial, sans-serif;


$u-button-primary: #0866C6;
$u-button-primary-active: #0753a1;
$u-button-secondary: #868e96;
$u-button-secondary-active: #767B84;
$u-button-positive: #21ba45;
$u-button-positive-active: #16ab39;
$u-button-negative: #dc3545;
$u-button-negative-active: #c82333;
$u-button-info: #17a2b8;
$u-button-info-active: #138496;
$u-button-dark: #343a40;
$u-button-dark-active: #23272b;

@mixin button-theme-modifier($color, $class) {
    &-#{$class} {
        background: $color;
        color: #fff;

        &:hover {
          background: darken($color, 5%);
        }
    }
}

.u-button {
    position: relative;
    display: inline-block;
    min-height: 1em;
    padding: $u-button-padding-y $u-button-padding-x;
    background-color: #e0e1e2;
    background-image: none;
    border: none;
    border-radius: $u-button-border-radius;
    box-shadow: $u-button-box-shadow;
    line-height: $u-button-line-height;
    vertical-align: baseline;
    color: rgba(0, 0, 0, 0.6);
    font-family: $u-button-font-family;
    font-weight: 400;
    font-size: 1rem;
    text-align: center;
    text-decoration: none;
    text-shadow: none;
    transition: color 0.3s ease-out, background-color 0.3s ease-out;

    outline: 0;
    user-select: none;
    cursor: pointer;

    will-change: '';
    -webkit-tap-highlight-color: transparent;

    &:hover,
    &:focus {
        background-color: #cacbcd;
    }

    &:not(:disabled) {
        cursor: pointer;
    }

    // theme
    @include button-theme-modifier($u-button-primary, "primary");
    @include button-theme-modifier($u-button-secondary, "secondary");
    @include button-theme-modifier($u-button-positive, "positive");
    @include button-theme-modifier($u-button-negative, "negative");
    @include button-theme-modifier($u-button-negative, "negative");
    @include button-theme-modifier($u-button-info, "info");
    @include button-theme-modifier($u-button-dark, "dark");
    

    &-tiny {
        padding: $u-button-padding-y-tiny $u-button-padding-x-tiny;
        font-size: $u-button-font-size-tiny;
    }

    &-small {
        padding: $u-button-padding-y-small $u-button-padding-x-small;
        font-size: $u-button-font-size-small;
    }

    &-large {
        padding: $u-button-padding-y-large $u-button-padding-x-large;
        font-size: $u-button-font-size-large;
    }

    &-big {
        padding: $u-button-padding-y-big $u-button-padding-x-big;
        font-size: $u-button-font-size-big;
    }
}
