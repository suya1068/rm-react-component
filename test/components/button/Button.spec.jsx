import React from "react";
import Button from "src/components/button/Button";

describe("Button Component", () => {
    it("랜더링한다.", () => {
        const component = shallow(<Button>버튼</Button>);
        expect(component).toMatchSnapshot();
    });

    it("테마를 설정한다.", () => {
        const component = shallow(<Button theme="primary">버튼</Button>);
        expect(component).toMatchSnapshot();

        component.setProps({ theme: "secondary" });
        expect(component).toMatchSnapshot();

        component.setProps({ theme: "positive" });
        expect(component).toMatchSnapshot();

        component.setProps({ theme: "negative" });
        expect(component).toMatchSnapshot();
    });
    
    it("사이즈를 설정한다.", () => {
        const component = shallow(<Button size="tiny">버튼</Button>);
        expect(component).toMatchSnapshot();

        component.setProps({ size: "small" });
        expect(component).toMatchSnapshot();

        component.setProps({ size: "large" });
        expect(component).toMatchSnapshot();

        component.setProps({ size: "big" });
        expect(component).toMatchSnapshot();
    });

    it("width와 height이 설정되었다면, inline-style로 데이터를 설정한다.", () => {
        let actual;

        actual = Button.createStyles();
        expect(actual).toEqual(null);

        actual = Button.createStyles({ width: 200, height: "" });
        expect(actual).toEqual({ width: "200px" });

        let component = shallow(<Button width={200}>버튼</Button>);
        expect(component).toMatchSnapshot();

        component = shallow(<Button width={200} height={50}>버튼</Button>);
        expect(component).toMatchSnapshot();
    });

    it("기본속성을 설정한다.", () => {
        const component = shallow(
            <Button
                className="text-center"
                autoFocus={true}
                style={{ width: "100px", backgroundColor: "black" }}
                onClick={() => {}}
            >버튼</Button>
        );
        expect(component).toMatchSnapshot();
    });

    it("width와 style속성 width과 중복된다면, width값이 설정된다.", () => {
        const component = shallow(<Button style={{ width: "100px" }} width={200}>버튼</Button>);
        expect(component).toMatchSnapshot();
    });
});

/*
    // https://semantic-ui.com/elements/button.html
    1. 유효하지 않은 테마를 설정하면, 타입에러가 발생한다.
    2. 유효하지 않은 사이즈를 설정하면, 타입에러가 발생한다.
    3. 버튼을 클릭하면, 콜백함수가 실행된다.

    // 기본 속성
    * type - one of ['button', 'submit']. default 'button'
    * disabled - one of [true, false]. default false
    * autofocus - one of [true, false]. default false

    // 커스텀 속성
    * theme - one of ['primary', 'secondary', 'positive', 'negative']. default null
    * size - one of ['tiny', 'small', 'large', 'big']. default null (? { width: number, height: number })
    * shape - one of ['circle', 'round', 'rectangle']. default 'round'
    * transparent - one of [true, false]. default false
*/

/*
    function Button() {}

    new Button();

    new Button({
        theme: 'primary'
    });

    new Button({
        theme: 'primary',
        size: 'small'
    });

    new Button({
        theme: 'primary',
        size: 'small',
        transparent: true
    });

    new Button({
        theme: 'primary',
        size: 'small',
        disabled: true
    });
*/
