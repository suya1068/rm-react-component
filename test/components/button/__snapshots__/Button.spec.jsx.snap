// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component width와 height이 설정되었다면, inline-style로 데이터를 설정한다. 1`] = `
<button
  className="u-button"
  styles={
    Object {
      "width": "200px",
    }
  }
>
  버튼
</button>
`;

exports[`Button Component width와 height이 설정되었다면, inline-style로 데이터를 설정한다. 2`] = `
<button
  className="u-button"
  styles={
    Object {
      "height": "50px",
      "width": "200px",
    }
  }
>
  버튼
</button>
`;

exports[`Button Component width와 style속성 width과 중복된다면, width값이 설정된다. 1`] = `
<button
  className="u-button"
  styles={
    Object {
      "width": "200px",
    }
  }
>
  버튼
</button>
`;

exports[`Button Component 기본속성을 설정한다. 1`] = `
<button
  autoFocus={true}
  className="u-button text-center"
  onClick={[Function]}
  styles={
    Object {
      "backgroundColor": "black",
      "width": "100px",
    }
  }
>
  버튼
</button>
`;

exports[`Button Component 랜더링한다. 1`] = `
<button
  className="u-button"
>
  버튼
</button>
`;

exports[`Button Component 사이즈를 설정한다. 1`] = `
<button
  className="u-button u-button-tiny"
>
  버튼
</button>
`;

exports[`Button Component 사이즈를 설정한다. 2`] = `
<button
  className="u-button u-button-small"
>
  버튼
</button>
`;

exports[`Button Component 사이즈를 설정한다. 3`] = `
<button
  className="u-button u-button-large"
>
  버튼
</button>
`;

exports[`Button Component 사이즈를 설정한다. 4`] = `
<button
  className="u-button u-button-big"
>
  버튼
</button>
`;

exports[`Button Component 테마를 설정한다. 1`] = `
<button
  className="u-button u-button-primary"
>
  버튼
</button>
`;

exports[`Button Component 테마를 설정한다. 2`] = `
<button
  className="u-button u-button-secondary"
>
  버튼
</button>
`;

exports[`Button Component 테마를 설정한다. 3`] = `
<button
  className="u-button u-button-positive"
>
  버튼
</button>
`;

exports[`Button Component 테마를 설정한다. 4`] = `
<button
  className="u-button u-button-negative"
>
  버튼
</button>
`;
